In [214]:
import pandas as pd

df = pd.read_excel('base_aed_evasao_pre_processamento.xlsx')

In [215]:
df.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho,Renda Familiar Total
0,18,feminino,Solteiro(a),1414.37,Pública,Ativo,NaN,Não,5,68.5,1414.37
1,20,FEM,Divorciado(a),4238.37,Privada,Evadido,Dificuldade financeira,Não,6,76.4,4238.37
2,18,Masculino,Viúvo(a),2295.80,Pública,Ativo,NaN,Não,9,77.8,2295.80
3,27,masc,Solteiro(a),3765.94,Pública,Evadido,Baixo desempenho,Sim,1,61.9,3765.94
4,30,Masculino,Solteir,3991.39,Privada,Ativo,NaN,Sim,1,61.6,3991.39


In [216]:
print(df.columns)
df.info()

Index(['Idade', 'Gênero', 'Estado Civil', 'Renda Familiar (R$)',
       'Escola Ensino Médio', 'Status', 'Motivo da Evasão',
       'Trabalha Atualmente', 'Último Período Cursado', 'Desempenho',
       'Renda Familiar Total'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Idade                   153 non-null    int64  
 1   Gênero                  153 non-null    object 
 2   Estado Civil            153 non-null    object 
 3   Renda Familiar (R$)     145 non-null    float64
 4   Escola Ensino Médio     145 non-null    object 
 5   Status                  153 non-null    object 
 6   Motivo da Evasão        51 non-null     object 
 7   Trabalha Atualmente     145 non-null    object 
 8   Último Período Cursado  153 non-null    int64  
 9   Desempenho              145 non-null    float64
 10  Renda

In [217]:
df_new = df[['Idade', 'Gênero', 'Estado Civil', 'Renda Familiar (R$)', 'Escola Ensino Médio', 'Status', 'Motivo da Evasão', 'Trabalha Atualmente', 'Último Período Cursado', 'Desempenho']]

# Atividade semana 4

## 1. Seleção dos dados

### 1.1 Análise inicial:
* Colunas e características:

| Coluna | Tipo | Propósito | Relevância |
|---|---|---|---|
| Idade | int64 | Informar a idade | Importante |
| Gênero | object | Informar o gênero | Importante |
| Estado Civil | object | Informar o estado civil | Importante |
| Renda Familiar (R$) | float64 | Informar a renda familiar | Importante |
| Escola Ensino Médio | object | Informar se a escola em que o aluno estudou no ensino médio é pública ou privada | Importante |
| Status | object | Informar se o aluno está ativo ou não | Importante |
| Motivo da Evasão | object | Informar o motivo da evasão | Importante |
| Trabalha Atualmente | object | Informar se o aluno trabalha | Importante |
| Último Período Cursado | int64 | Informar o último período cursado | Importante |
| Desempenho | float64 | Informar o desempenho do aluno | Importante |
| Renda Familiar Total | float64 | Informar a renda familiar | Irrelevante — duplicata da coluna "Renda Familiar (R$)" |

### 1.2 Identificação de redundâncias:
* A coluna 'Renda Familiar Total' é redundante pois contém os mesmos valores da coluna 'Renda Familiar (R$)'.
* Uma coluna redundante pode afetar a análise desejada devido ao impacto em cálculos e desempenho de processamento.

### 1.3 Seleção final:
* As variáveis que permanecerão no DF são:
    - Idade
    - Gênero
    - Estado Civil
    - Renda Familiar (R$)
    - Escola Ensino Médio
    - Status
    - Motivo da Evasão
    - Trabalha Atualmente
    - Último Período Cursado
    - Desempenho
    

## 2. Pré-processamento

### 2.1 Limpeza de dados:

#### 2.1.1 Registros duplicados:

Vamos verificar se há linhas duplicadas

In [218]:
duplicados = df_new.duplicated(keep='first')

print(f'Linhas duplicadas: {duplicados.sum()}')

df_new[duplicados]

Linhas duplicadas: 3


,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
150,27,Masculino,Viúvo(a),4892.37,Pública,Ativo,NaN,Sim,6,69.4
151,20,Feminino,Divorciado(a),2912.31,Pública,Evadido,Mudança de curso,Sim,6,71.2
152,17,Feminino,Viúvo(a),899.56,Privada,Evadido,Baixo desempenho,Não,6,77.6


Visualizando ambas as linhas repetidas:

In [219]:
df_new[df_new.duplicated(keep=False)]

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
14,27,Masculino,Viúvo(a),4892.37,Pública,Ativo,NaN,Sim,6,69.4
75,17,Feminino,Viúvo(a),899.56,Privada,Evadido,Baixo desempenho,Não,6,77.6
98,20,Feminino,Divorciado(a),2912.31,Pública,Evadido,Mudança de curso,Sim,6,71.2
150,27,Masculino,Viúvo(a),4892.37,Pública,Ativo,NaN,Sim,6,69.4
151,20,Feminino,Divorciado(a),2912.31,Pública,Evadido,Mudança de curso,Sim,6,71.2
152,17,Feminino,Viúvo(a),899.56,Privada,Evadido,Baixo desempenho,Não,6,77.6


Excluindo os primeiros registros duplicados e verificando se o procedimento foi realizado com sucesso: 

In [220]:
df_sem_duplicadas = df_new.drop_duplicates().copy()
print(f'Linhas duplicadas: {df_sem_duplicadas.duplicated().sum()}')

Linhas duplicadas: 0


#### 2.1.2 Valores nulos:

Primeiramente realizamos uma verificação da quantidade total de linhas do DF completo através da função '.shape', e em seguida apresentamos a quantidade de valores nulos em cada coluna.

Obs.: tentei utilizar o .describe() da primeira vez, mas notei que ele não trazia as infos de colunas não numéricas. Deste modo, optei por contar os valores nulos em cada coluna.

In [221]:
print(f'Shape do DF sem duplicadas: {df_sem_duplicadas.shape[0]}')

nulos = df_sem_duplicadas.isna().sum()
print(f'\nQuantidade de valores nulos em cada coluna:\n{nulos}')

Shape do DF sem duplicadas: 150

Quantidade de valores nulos em cada coluna:
Idade                       0
Gênero                      0
Estado Civil                0
Renda Familiar (R$)         8
Escola Ensino Médio         8
Status                      0
Motivo da Evasão          101
Trabalha Atualmente         8
Último Período Cursado      0
Desempenho                  8
dtype: int64


Notamos que as colunas 'Renda Familiar ($)', 'Escola Ensino Médio', 'Trabalha Atualmente' e 'Desempenho' possuem a mesma de quantidade de valores nulos: 8.

Outro ponto interessante é que a coluna 'Motivo da Evasão' apresenta 101 valores nulos (alunos que estão ativos ou se formaram), ou seja, 49 alunos que evadiram (150 - 101 = 49).

Para validar se esse número realmente representa a quantidade de alunos evadidos e garantir que não haja valores não preenchidos, realizamos uma verificação da quantidade de alunos evadidos baseado no shape do DF:

In [222]:
evasao = df_sem_duplicadas[df_sem_duplicadas['Status'] == 'Evadido']
print(f'\nShape da quantidade de alunos evadidos: {evasao.shape[0]}')


Shape da quantidade de alunos evadidos: 54


Notamos uma discrepância entre os valores, onde anteriormente tínhamos o valor de 49 alunos evadidos, agora percebemos que há na verdade 54 alunos evadidos.

Isso indica que há 5 alunos evadidos com a coluna 'Motivo da Evasão' não preenchida.

Vamos validar filtrando pelos alunos evadidos e com a coluna em questão com valores nulos:

In [223]:
evadidos_sem_motivo = df_sem_duplicadas[(df_sem_duplicadas['Status'] == 'Evadido') & (df_sem_duplicadas['Motivo da Evasão'].isna())]
print(f'\nQuantidade de registros evadidos sem motivo de evasão (nulo): {evadidos_sem_motivo.shape[0]}')


Quantidade de registros evadidos sem motivo de evasão (nulo): 5


Vamos substituir esses valores nulos por 'Não Informado' para não introduzir viés em nosso DF, e em seguida validar se os valores nulos para o status evadido foram zerados.

In [224]:
filtro = (df_sem_duplicadas['Status'] == 'Evadido') & (df_sem_duplicadas['Motivo da Evasão'].isna())

df_sem_duplicadas.loc[filtro, 'Motivo da Evasão'] = 'Não Informado'

evadidos_sem_motivo = df_sem_duplicadas[(df_sem_duplicadas['Status'] == 'Evadido') & (df_sem_duplicadas['Motivo da Evasão'].isna())]
print(f'\nQuantidade de registros evadidos sem motivo de evasão (nulo): {evadidos_sem_motivo.shape[0]}')


Quantidade de registros evadidos sem motivo de evasão (nulo): 0


Vamos agora verificar as outras colunas com valores nulos:

In [225]:
colunas_nulo = df_sem_duplicadas[
    (df_sem_duplicadas['Renda Familiar (R$)'].isna()) +
    (df_sem_duplicadas['Escola Ensino Médio'].isna()) +
    (df_sem_duplicadas['Trabalha Atualmente'].isna()) +
    (df_sem_duplicadas['Desempenho'].isna())]

qtd_linhas = colunas_nulo.shape[0]
porcentagem = int(colunas_nulo.shape[0]) / int(df_sem_duplicadas.shape[0]) * 100

print(f'Quantidade de linhas onde alguma das colunas\n(Renda Familiar (R$),\nEscola Ensino Médio,\nTrabalha Atualmente,\nDesempenho)\ntem valor nulo: {qtd_linhas}')
print(f'Porcentagem de {qtd_linhas} linhas em relação ao DF completo: {porcentagem}')

colunas_nulo

Quantidade de linhas onde alguma das colunas
(Renda Familiar (R$),
Escola Ensino Médio,
Trabalha Atualmente,
Desempenho)
tem valor nulo: 29
Porcentagem de 29 linhas em relação ao DF completo: 19.333333333333332


,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
10,33,Feminino,Casado(a),2798.23,Privada,Evadido,Baixo desempenho,NaN,6,150.0
18,34,Feminino,Viúvo(a),NaN,Privada,Ativo,NaN,Não,1,62.0
20,25,Feminino,Solteiro(a),1013.70,Privada,Evadido,Trabalho,Não,7,NaN
30,18,Outro,Solteiro(a),2112.46,Pública,Evadido,Baixo desempenho,NaN,1,62.2
31,28,Feminino,Solteiro(a),NaN,Privada,Evadido,Baixo desempenho,Não,7,77.9
45,29,Masculino,Solteiro(a),1743.56,Privada,Evadido,Trabalho,Não,9,NaN
47,26,Masculino,Casado(a),1008.10,Privada,Formado,NaN,Não,1,NaN
51,33,Outro,Viúvo(a),2259.32,NaN,Formado,NaN,Não,1,61.9
54,21,Feminino,Casado(a),1689.11,Privada,Evadido,Trabalho,NaN,10,82.7
58,17,Feminino,Divorciado(a),3525.10,Pública,Ativo,NaN,NaN,1,62.6


29 linhas e 19,33% do DF completo é bastante coisa, então não podemos remover as linhas. Vamos substituir os valores. Podemos notar na coluna Desempenho que o valor máximo está sendo apresentado como 150, o que indica um outlier e um erro, tendo em vista que o desempenho é baseado em valores de 0 a 100.

In [226]:
df_sem_duplicadas['Desempenho'].describe()

count    142.000000
mean      73.216197
std        9.813817
min       61.500000
25%       66.700000
50%       72.200000
75%       78.550000
max      150.000000
Name: Desempenho, dtype: float64

Deste modo, vamos calcular o valor da mediana sem considerar esse valor de 150, transformar esse valor em nulo, em seguida imputar o valor da mediana nessa posição, e então validar se está devidamente ajustado.

In [227]:
df_sem_duplicadas.loc[df_sem_duplicadas['Desempenho'] > 100, 'Desempenho'] = None

mediana_desempenho = df_sem_duplicadas['Desempenho'].median()

df_sem_duplicadas['Desempenho'] = df_sem_duplicadas['Desempenho'].fillna(mediana_desempenho)

df_sem_duplicadas.describe()

,Idade,Renda Familiar (R$),Último Período Cursado,Desempenho
count,150.000000,142.000000,150.000000,150.000000
mean,24.506667,2887.906056,5.553333,72.643333
std,5.184834,4112.962307,2.968598,7.162621
min,17.000000,500.000000,1.000000,61.500000
25%,20.000000,1833.075000,3.000000,67.025000
50%,24.000000,2468.720000,6.000000,72.200000
75%,29.000000,3261.365000,8.000000,77.800000
max,35.000000,50000.000000,10.000000,88.700000


Show de bola! Vamos para a próxima coluna: Renda Familiar (R$).

Nesta coluna, vamos realizar o mesmo procedimento realizado na coluna Desempenho, tendo em vista o valor outlier de 50000.

Com isso, vamos calcular o valor da mediana sem considerar esse valor de 500000, transformar esse valor em nulo, em seguida imputar o valor da mediana nessa posição, e então validar se está devidamente ajustado.

In [228]:
print(df_sem_duplicadas['Renda Familiar (R$)'].sort_values(ascending=False))

5      50000.00
14      4892.37
11      4861.60
82      4843.47
65      4700.70
         ...   
73          NaN
76          NaN
78          NaN
118         NaN
141         NaN
Name: Renda Familiar (R$), Length: 150, dtype: float64


In [229]:
df_sem_duplicadas.loc[df_sem_duplicadas['Renda Familiar (R$)'] > 4900, 'Renda Familiar (R$)'] = None

mediana_renda = df_sem_duplicadas['Renda Familiar (R$)'].median()

df_sem_duplicadas['Renda Familiar (R$)'] = df_sem_duplicadas['Renda Familiar (R$)'].fillna(mediana_renda)

df_sem_duplicadas['Renda Familiar (R$)'].describe()

count     150.000000
mean     2548.529667
std      1003.331879
min       500.000000
25%      1940.707500
50%      2466.310000
75%      3175.932500
max      4892.370000
Name: Renda Familiar (R$), dtype: float64

Vamos para a próxima coluna: Escola Ensino Médio.

Nesta situação, para sermos conservadores, vamos alterar os valores nulos para "Desconhecida".

In [230]:
df_sem_duplicadas['Escola Ensino Médio'] = df_sem_duplicadas['Escola Ensino Médio'].fillna('Desconhecida')

print(df_sem_duplicadas['Escola Ensino Médio'].isna().sum())

0


Última coluna a ter os valores nulos tratados: Trabalha Atualmente

Vamos realizar o mesmo processo da coluna 'Escola Ensino Médio', imputando 'Desconhecido' aos valores nulos.

In [231]:
df_sem_duplicadas['Trabalha Atualmente'] = df_sem_duplicadas['Trabalha Atualmente'].fillna('Desconhecido')

print(df_sem_duplicadas['Trabalha Atualmente'].isna().sum())

0


Podemos observar que agora não temos mais valores nulos (desconsiderando os 101 registros nulos de Motivos da Evasão, que representam alunos ativos ou formados):

In [232]:
df_tratado = df_sem_duplicadas

df_tratado.isna().sum()

Idade                      0
Gênero                     0
Estado Civil               0
Renda Familiar (R$)        0
Escola Ensino Médio        0
Status                     0
Motivo da Evasão          96
Trabalha Atualmente        0
Último Período Cursado     0
Desempenho                 0
dtype: int64

### 2.2 Integração dos dados:

Caso houvesse outra planilha para ser integrada a essa em questão, isso não seria possível, tendo em vista a falta de uma chave primária, ou de um conjunto de colunas que poderiam identificar um aluno.

De qualquer modo, alguns cuidados devem ser adotados ao realizar uma integração entre tabelas:

* Não pode haver chaves duplicadas ou nulas;
* Tipos de dados diferentes entre as chaves das tabelas. Exemplo: 'id' em uma tabela é int e em outra é string...

### 2.3 Redução de dados:

A única coluna removida foi Renda Familiar Total, por ser redundante. As demais colunas foram mantidas pois possuem variação adequada e relevância analítica. A coluna Motivo da Evasão, apesar de 101 nulos, foi preservada pois esses valores representam alunos não evadidos, sendo nulos semanticamente justificados.

### 2.4 Transformação de dados:

#### 2.4.1 Padronizando valores de texto:

Ao visualizar as primeiras linhas do DF, podemos observar uma inconsistência no padrão dos valores da coluna 'Gênero':

In [233]:
df_tratado.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
0,18,feminino,Solteiro(a),1414.37,Pública,Ativo,NaN,Não,5,68.5
1,20,FEM,Divorciado(a),4238.37,Privada,Evadido,Dificuldade financeira,Não,6,76.4
2,18,Masculino,Viúvo(a),2295.80,Pública,Ativo,NaN,Não,9,77.8
3,27,masc,Solteiro(a),3765.94,Pública,Evadido,Baixo desempenho,Sim,1,61.9
4,30,Masculino,Solteir,3991.39,Privada,Ativo,NaN,Sim,1,61.6


Vamos validar em quais colunas a mesma inconsistência ocorre:

Notamos que as colunas 'Gênero' e 'Estado Civil' possuem mais de uma palavra para um mesmo dado. Com isso, vamos padronizar e organizar estes dados, validando se os ajustes ocorreram corretamente.

In [234]:
for col in df_tratado.columns:
    print(df_tratado[col].value_counts())
    print()

Idade
21    13
20    12
18    12
24    12
17    11
25    10
33    10
29     9
27     8
23     8
30     8
22     8
19     6
28     5
34     5
26     4
32     4
31     4
35     1
Name: count, dtype: int64

Gênero
Feminino     57
Masculino    47
Outro        43
feminino      1
FEM           1
masc          1
Name: count, dtype: int64

Estado Civil
Casado(a)        46
Viúvo(a)         39
Divorciado(a)    34
Solteiro(a)      27
Solteir           1
Casada            1
Viuvo             1
Separado          1
Name: count, dtype: int64

Renda Familiar (R$)
2466.31    10
500.00      2
2295.80     1
1414.37     1
3765.94     1
           ..
4169.38     1
2966.37     1
2471.13     1
3141.08     1
722.38      1
Name: count, Length: 140, dtype: int64

Escola Ensino Médio
Privada         78
Pública         64
Desconhecida     8
Name: count, dtype: int64

Status
Ativo      65
Evadido    54
Formado    31
Name: count, dtype: int64

Motivo da Evasão
Baixo desempenho          18
Dificuldade financeira    

In [235]:
#Gênero
#Feminino     57
#Masculino    47
#Outro        43
#feminino      1
#FEM           1
#masc          1

df_tratado['Gênero'] = df_tratado['Gênero'].replace(['feminino', 'FEM'], 'Feminino')
df_tratado['Gênero'] = df_tratado['Gênero'].replace('masc', 'Masculino')

#Estado Civil
#Casado(a)        46
#Viúvo(a)         39
#Divorciado(a)    34
#Solteiro(a)      27
#Solteir           1
#Casada            1
#Viuvo             1
#Separado          1

df_tratado['Estado Civil'] = df_tratado['Estado Civil'].replace('Solteir', 'Solteiro(a)')
df_tratado['Estado Civil'] = df_tratado['Estado Civil'].replace('Casada', 'Casado(a)')
df_tratado['Estado Civil'] = df_tratado['Estado Civil'].replace('Viuvo', 'Viúvo(a)')
df_tratado['Estado Civil'] = df_tratado['Estado Civil'].replace('Separado', 'Divorciado(a)')


print(df_tratado['Gênero'].value_counts())
print()
print(df_tratado['Estado Civil'].value_counts())

Gênero
Feminino     59
Masculino    48
Outro        43
Name: count, dtype: int64

Estado Civil
Casado(a)        47
Viúvo(a)         40
Divorciado(a)    35
Solteiro(a)      28
Name: count, dtype: int64


#### 2.4.2 Conversão de variáveis categóricas para números

In [236]:
mapeamento = {
    'Gênero':               {'Feminino': 1, 'Masculino': 2, 'Outro': 3},
    'Estado Civil':         {'Solteiro(a)': 1, 'Divorciado(a)': 2, 'Viúvo(a)': 3, 'Casado(a)': 4},
    'Escola Ensino Médio':  {'Pública': 1, 'Privada': 2, 'Desconhecida': 3},
    'Status':               {'Ativo': 1, 'Evadido': 2, 'Formado': 3},
    'Trabalha Atualmente':  {'Não': 1, 'Sim': 2, 'Desconhecido': 3} 
}

for col, mapa in mapeamento.items():
    df_tratado[col] = df_tratado[col].map(mapa)

df_tratado.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
0,18,1,1,1414.37,1,1,NaN,1,5,68.5
1,20,1,2,4238.37,2,2,Dificuldade financeira,1,6,76.4
2,18,2,3,2295.80,1,1,NaN,1,9,77.8
3,27,2,1,3765.94,1,2,Baixo desempenho,2,1,61.9
4,30,2,1,3991.39,2,1,NaN,2,1,61.6


### 2.4.3 Criação de variáveis derivadas

In [237]:
for i in df_tratado.index:
    idade = df_tratado.loc[i, 'Idade']

    if idade <= 21:
        df_tratado.loc[i, 'Faixa Etária'] = 'Jovem'
    elif idade <= 60:
        df_tratado.loc[i, 'Faixa Etária'] = 'Adulto'
    else:
        df_tratado.loc[i, 'Faixa Etária'] = 'Maduro'

In [238]:
min_renda = df_tratado['Renda Familiar (R$)'].min()
max_renda = df_tratado['Renda Familiar (R$)'].max()

print(min_renda, max_renda)

df_tratado['Renda Familiar Normalizada'] = (df_tratado['Renda Familiar (R$)'] - min_renda) / (max_renda - min_renda)

500.0 4892.37


In [239]:
df_tratado.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho,Faixa Etária,Renda Familiar Normalizada
0,18,1,1,1414.37,1,1,NaN,1,5,68.5,Jovem,0.208172
1,20,1,2,4238.37,2,2,Dificuldade financeira,1,6,76.4,Jovem,0.851105
2,18,2,3,2295.80,1,1,NaN,1,9,77.8,Jovem,0.408845
3,27,2,1,3765.94,1,2,Baixo desempenho,2,1,61.9,Adulto,0.743548
4,30,2,1,3991.39,2,1,NaN,2,1,61.6,Adulto,0.794876


### 2.4.4 Descrição de melhorias das transformações:

Todas essas etapas de transformações foram essenciais para padronizar e simplificar a análise geral do DF. Deste modo, deixando o DF pronto para análises mais profundas e elaboradas.

## 3. Exportação do DataFrame como planilha Excel:

In [240]:
df_tratado.to_excel('base_aed_evasao_pre_processamento_tratada.xlsx', index=False, sheet_name='Dados Tratados')